# Errores en métodos numéricos

## Definiciones y estándares con respecto a los errores y precisión numérica

Para medir los errores numéricos usamos la nomenclatura de los errores experimentales. Por ejemplo, usamos la notación de dar el valor resultante $\pm$ el error.


In [ ]:
from random import *
l = []
for i in range(10):
    x = uniform(0.8, 2.8)
    l.append(x)

n = len(l)
suma = 0
for i in range(n):
    suma += l[i]
    promedio = suma/n

error = max(l) - min(l)
print('La variable pseudoaleatoria resultante es: \n', promedio, '+-', error)

La variable pseudoaleatoria resultante es: 
 1.6453652307178863 +- 1.670069138444813


### Error absoluto, error relativo y error porcentual

Usamos $X_{teo}$ para el valor teórico o verdadero de alguna variable y $X_{aprox}$ para el valor aproximado o numérico.

El error absoluto está definido por:

$$e_{abs}(X_{aprox})=|X_{teo}-X_{aprox}|$$

El error relativo es:
$$e_{rel}(X_{aprox})=\frac{|X_{teo}-X_{aprox}|}{X_{teo}}$$

El error porcentual es:
$$e_{por}(X_{aprox})=e_{rel}(X_{aprox})(100\%)$$

### Ejemplo:
Sea $X_{teo}=e$ y $X_{aprox}=\tfrac{19}{7}$

$$e_{abs}(X_{aprox})=|e - \frac{19}{7}| \approx 0.003996$$

El error relativo es:
$$e_{rel}(X_{aprox})=\frac{|e - \frac{19}{7}|}{e} \approx 0.00147$$

El error porcentual es:
$$e_{por}(X_{aprox})=(0.00147)(100\%)=1.47\%$$


In [ ]:
from math import *
def error_abs(real,approx):
    return abs(real-approx)
def error_rel(real,approx):
    return (abs(real-approx))/real
def error_por(real,approx):
    return ((abs(real-approx))/real)*100

x_teo=e
x_aprox=19/7

e_abs=error_abs(x_teo,x_aprox)
e_rel=error_rel(x_teo,x_aprox)
e_por=error_por(x_teo,x_aprox)

print('El error absoluto es : %5.4f' %e_abs) #5 digitos donde hay 4 decimales (4f)
print('El error relativo es : %5.4f' %e_rel)
print('El error porcentual es : %5.7f' %e_por)

El error absoluto es : 0.0040
El error relativo es : 0.0015
El error porcentual es : 0.1470088


## Digitos significativos

### Definición:
Decimos que $X_{aprox}$ tiene $m$ digitos significativos cuando el error de $X_A$ cumple con $err(X_{aprox})\leq tol$ en el digito (m+1) distintos de cero.

### Ejemplo
$$e_{abs}(X_{aprox})=\left|e-\frac{19}{7}\right|\approx 0.00\mathbf{3}996$$
se tiene 3 digitos significativos

## Tipos de errores
1. Errores numéricos derivados de la aproximación del método numérico $tol$
2. Errores numéricos de la aritmética discreta de la computadora $X_{aprox}$.
3. Errores en el planteamiento del modelo (¡Vacas esféricas, cuidado!)

## 1. Errores numéricos de la aproximación

Aquí consideramos los términos que despreciamos de las aproximaciones numéricos.

Serie de Taylor
$$e^x \approx 1 + x + \frac{1}{2} x^2$$

Diferenciación
$$ f'(x) \approx \frac{f(x + h) - f(x - h)}{2h}$$

Integración numérica
$$ \int_0^1 f(x) dx \approx \frac{1}{n} \sum_{j=1}^n f(\frac{j}{n})$$

Donde definimos el error total como:
$$ e_{T} = e_n + e_b$$

Donde:

$e_n$ = error numerico

$e_b$=error de bits



## 2. Errores numéricos de la aritmética discreta

Hay que considerar que la computadora tiene limitaciones. Las computadoras no pueden guardar números con decimales infinitos, por lo que no se puede manipular elementos continuos, sino que toda la matemáticas que se hagan con la computadora serán discretas. Tampoco números muy grandes o muy chicos.

In [ ]:
x=1.1+2.2
if x==3.3:
    print(x)
else:
    print('x distinto a 3.3')
#¿Por qué no se imprimió x terminando el "if"? Ocurre que por error de redonde
#1.1+2.2 no es exactamente 3.3, sino que es 3.300000000000003. Por lo cual no
#se cumple la condición del if. Esto es bastante común en física computacional,
#por lo que la mejor manera de tratar esto, es introducir un valor
#epsilón en el código.
print(x)

x distinto a 3.3
3.3000000000000003


In [ ]:
x=1.1+2.2
epsilon=1e-12
if abs(x-3.3)<epsilon: #si el valor absuluto() es menor a epsilon, entonces... #abs=valor absoluto
    print(x)

3.3000000000000003


En Python, el nivel de precisión estandar es de 16 cifras significativas. Esto significa que los números irracionales serán truncados:

Valor verdadero de Pi: 3.1415956535897932384626...

Valor en Python de Pi: 3.141595653589793

Diferencias:           0.0000000000000002384626

A este error se le llama error de redondeo.

Entonces, el error numérico de un número, que se denotará $\epsilon$ (epsilon), se define como la cantidad calculada por la computadora menos el valor real (el error absoluto).

En general, si $X_{aprox}$ es una aproximación a cierto número de cifras significativas, digamos 16, entonces el error de redondeo tiene un valor típico de $\frac{X_{aprox}}{10^{16}}$. Usualmente es una buena aproximación considerar el error como una cantidad uniformemente distribuida de un número con la desviación estandar $\sigma=CX_{aprox}$, donde $C\approx 10^{-16}$. Nos referimos a $C$ como una constante de error.

Comúnmente nos referimos a $\epsilon$ como la desviación estandar ya que sí dieramos un valor exacto de $\epsilon$ sería muy fácil obtener el valor real de x.

Veamos con más cuidado esto de los errores y la precisión de los cálculos:

Sean:
        x=1000000000000000
        y=1000000000000001.234567890123478
Recordando que la computadora sólo guarda 16 cifras significativas, y será truncada. La diferencia entre y-x=1.2, cuando en realidad es 1.234567890123478, lo cual el error porcentual es del 0.23%. Sin embargo, al agregar una cifra decimal a x y y, el error porcentual se incrementa considerablemente.

Veamos en código como se comporta todo.

In [ ]:
x = 10000000000000000
y = 10000000000000001.234567890123478
a = y - x
error_real = 1.234567890123478
print(a)
#error = (abs(a)/(error_real))*100
error = (abs(a-error_real)/(error_real))*100
print(error, '%')

2.0
62.000001457997236 %


In [ ]:
#Primero, veremos el paquete decimal
from decimal import *
print(1/7.)

getcontext().prec = 6 #con "prec" nos referimos a la precision, ie, el numero de decimales que queremos nos devuelva
print(Decimal(1) / Decimal(7))

getcontext().prec = 100000 #queremos una presicion de 100000 decimales de la siguiente operacion:
print(Decimal(1) / Decimal(7))

0.14285714285714285
0.142857
0.142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142857142

### Ejemplo

Evaluamos
$$f(x) = x(\sqrt{x + 1} - \sqrt{x})$$
con una calculadora de 6 digitos de precisión.

In [ ]:
from math import sqrt
from decimal import *
x = 100000001111

getcontext().prec = 4
print(Decimal(x*(sqrt(x+1) - sqrt(x))))

158115.29186200373806059360504150390625


In [ ]:
from IPython.display import Image
Image("Errores_1.png")

In [ ]:
#### Paréntesis de Pandas ####

from numpy import sqrt
import pandas
tabla = {'A': [1, 2, 3], 'B': [9, 20, 22], 'C': [54, 12, 13]}
df = pandas.DataFrame(tabla)
df.apply(sqrt)

A         B         C
0  1.000000  3.000000  7.348469
1  1.414214  4.472136  3.464102
2  1.732051  4.690416  3.605551

In [ ]:
from decimal import Decimal as D, getcontext as C #C=getcontext()
import pandas as pd
from math import sqrt

C().prec = 6

def f(x):
    return x * (sqrt(x + 1) - sqrt(x))
def fd(x):
    return x * (C().sqrt(x + D(1)) - C().sqrt(x))
residuo = []
idx = []
for i in range(10):
    x = 10**i
    idx.append(x)
    residuo.append([fd(D(x)), f(x)])
#print(residuo)
cols = ["Aprox", "Real"]
pd.DataFrame(residuo, index=idx, columns=cols)

Aprox          Real
1           0.41421      0.414214
10          1.54340      1.543471
100          4.9900      4.987562
1000        15.8000     15.807437
10000        50.000     49.998750
100000      100.000    158.113488
1000000           0    499.999875
10000000       0.00   1581.138790
100000000         0   5000.000056
1000000000      0.0  15811.390767

### Ejemplo

Definimos
$$f_1(x) = \frac{1 - \cos(x)}{x^2}, \qquad x \not= 0$$
evaluamos con 10 digitos de precisión

In [ ]:
from decimal import Decimal as D, getcontext as C #C=getcontext()
import pandas as pd
from math import sqrt, cos

C().prec = 10
#from math import cos
def f1(x):
    return (1 - cos(x)) / (x*x)
def f1d(x):
    return +D(D(1) - (+D(cos(x)))) / (D(x*x))
residuo = []
idx = []
for i in range(1, 7):
    x = 10**(-i)
    idx.append(x)
    err=abs(float(f1d(+D(x)))-f1(x))
    residuo.append([f1d(+D(x)), f1(x),err*100])
pd.set_option("precision", C().prec)
cols = ["Aprox", "Real","%"]
pd.DataFrame(residuo, index=idx, columns=cols)

Aprox          Real              %
0.100000  0.49958347  0.4995834722   0.0000002197
0.010000    0.499996  0.4999958333   0.0000166653
0.001000         0.5  0.4999999583   0.0000041674
0.000100         0.5  0.4999999970   0.0000003039
0.000010           1  0.5000000414  49.9999958630
0.000001       0E+12  0.5000444503  50.0044450291

In [ ]:
x = +D(0.001)
+D(cos(x))  #Redondeamos el resultado final regresando a la precisión default

Decimal('0.9999995000000416744967424165')

### 3. Errores en el planteamiento

Es consecuencia de restar dos numeros casí iguales, generan perdidas de significancia. Es difícil de detectar y no siempre es fácil evitarlo.

###  Ejemplo:
Sea $f(x) = \frac{1 - \cos(x)}{x^2}$, usando $$\cos(2\theta) = 2\cos^2(\theta) - 1 = 1 - 2\sin^2(\theta)$$

Con $x=2\theta$:
\begin{eqnarray}
f(x) &=& \frac{1 - \cos(x)}{x^2} = \frac{2\sin^2(x/2)}{x^2}\\
     &=& \frac{1}{2} \left[\frac{\sin(x/2)}{x/2}\right]^2
\end{eqnarray}

Comparar varios valores muy pequeños $x\approx 0.001$

In [ ]:
from decimal import Decimal as D, getcontext as C
from math import sin

C().prec = 10

def f1(x):
    return (1 - cos(x)) / (x*x)
def f1d(x):
    return +D(D(1) - (+D(cos(x)))) / (D(x*x))

def f1a(x):
    return 0.5 * (sin(x/2) / (x/2))**2
def f1da(x):
    return +D(0.5) * (+D(sin(x/2))/ (x/2))**2

#f1 sin aproximar y con 16 digitos
#f1a sin aproximar y con 16 digitos
#f1d con aproximacion y con 10 digitos

#f1da con aproximacionr y con 10 digitos

print('Funcion original y 16 digitos',+D(f1(0.00001)))
print('Funcion aprox y 16 digitos',+D(f1a(0.00001)))
print('Funcion original y 10 digitos',f1d(+D(0.00001)))
print('Funcion aprox y 10 digitos',f1da(+D(0.00001)))

Funcion original y 16 digitos 0.5000000414
Funcion aprox y 16 digitos 0.5000000000
Funcion original y 10 digitos 1
Funcion aprox y 10 digitos 0.500


Notamos que el resultado approximado y el correcto están mucho más cerca, el error casi desapareció.

In [ ]:
res = []
idx = []
for i in range(1, 7):
    x = 10**(-i)
    idx.append(x)
    res.append([f1d(+D(x)), f1da(+D(x)), f1(x), f1a(x)])
cols = ["Apr cos", "Apr sen", "Exacto cos", "Exacto sen"]
pd.set_option("precision", C().prec)
pd.DataFrame(res, index=idx, columns=cols)


Apr cos  \
0.1000000000000000055511151231  0.4995834721974178882960035747   
0.0100000000000000002081668171  0.4999958333473663932291007598   
0.0010000000000000000208166817  0.4999999583255032367668200230   
0.0001000000000000000047921736  0.4999999969612644666182642672   
0.0000100000000000000008180305  0.5000000413701854181969393176   
0.0000009999999999999999547481  0.5000444502911705452559110933   

                                                       Apr sen  \
0.1000000000000000055511151231  0.4995834721974233753935005396   
0.0100000000000000002081668171  0.4999958333472221389609766134   
0.0010000000000000000208166817  0.4999999583333346570012334245   
0.0001000000000000000047921736  0.4999999995833333802601472426   
0.0000100000000000000008180305  0.4999999999958332755258753269   
0.0000009999999999999999547481  0.4999999999999582836273477264   

                                                    Exacto cos  \
0.1000000000000000055511151231  0.4995834721974178327386084675   
0.0100000000000000002081668171  0.4999958333473664140456094174   
0.0010000000000000000208166817  0.4999999583255032575834775344   
0.0001000000000000000047921736  0.4999999969612645145389251411   
0.0000100000000000000008180305  0.5000000413701853885228842955   
0.0000009999999999999999547481  0.5000444502911705058068037033   

                                                    Exacto sen  
0.1000000000000000055511151231  0.4995834721974234393648828245  
0.0100000000000000002081668171  0.4999958333472221405635593783  
0.0010000000000000000208166817  0.4999999583333346597768809261  
0.0001000000000000000047921736  0.4999999995833334098804812129  
0.0000100000000000000008180305  0.4999999999958332219662793250  
0.0000009999999999999999547481  0.4999999999999582556142740941

### Ejemplo (más sutil)

Evaluamos $e^{-5}$ usando la aproximación de Taylor
$$e^{-5} \approx 1 + \frac{(-5)}{1!} + \frac{(-5)^2}{2!} + \frac{(-5)^3}{3!} + \dots + \frac{(-5)^n}{n!}$$

Usando $n = 25$, el error es de
$$\left| \frac{(-5)^{26}}{26!} e^c \right| \le 10^{-8}$$
Lo valuamos con 4 digitos de precisión.

In [ ]:
from math import exp, factorial
C().prec = 4
res = +D(1)
for i in range(1, 26):
    res = res + (+D((-5)**i/factorial(i)))

(res, +D(exp(-5)), +D(exp(-5)) - res)

(Decimal('0.009985'), Decimal('0.006738'), Decimal('-0.003247'))

Notamos la diferencia grande entre el resultado correcto 0.006738 y la aproximación 0.009985, con un error de -0.003247.

Para entender el efecto analizamos el termino
$$\frac{(-5)^3}{3!} = - \frac{125}{6} \approx -20.83$$

In [ ]:
term = (-5)**3 / factorial(3)
aprox = +D(term)
error = +D(term - float(aprox))
(term, aprox, error)

(-20.833333333333332, Decimal('-20.83333333333333214909544040'), Decimal('0'))

Veamos que el error de -0.003333 es del mismo orden de magnitud que el resultado fínal de 0.006738.

**Sumar (restar) terminos con un resultado mucho menor que los terminos individuales resulta en una perdida importante de significancia.**

## Underflow y Overflow
Python permite tener números lo suficientemente grandes o chicos, alrededor de 16 cifras significativas. Por ejemplo, el valor flotante más largo que se puede ingresar es $10^{308}$. Este número suele ser suficiente para hacer física computacional, pero en ocasiones, no es suficiente. Si nos pasamos de este número, ocurre que la variable se sobresatura "overflowed" y Python nos regresará un valor "inf", llevando a errores numéricos graves o a que el código no pueda ser ejecutado.

(Nota: Para números muy pequeños, como $2^{-1022}$)

In [ ]:
def factorial_entera(n):
    multiplicacion=1 #Funcion entera
    for i in range(1,n):
        multiplicacion*=i
    return multiplicacion
def factorial_flotante(n):
    multiplicacion=1.0 #Funcion flotante
    for i in range(1,n):
        multiplicacion*=i
    return multiplicacion

print('*El factorial entero es :\n',factorial_entera(200000))
print('**El factorial flotante es :\n',factorial_flotante(200))

*El factorial entero es :
 7101126727351572024834731668411529880449826783732008113481123723146133892580498428250412767039395406648965676076880220395780174977283962201494538491635785435331431515912508811609542030628057286905238189858996756360648473225155983473144301800814278458162232438519474018912580140997789407905893439707954869671777596266892974442997785094510774488507446495276544492489978186542791618812361701486489928843076919219088838087411682904441604153389238693036397400990971077222685423955446142115436605968376100901400164614271248782780548690443339125750680785294599831760924811636628872831507970440799337972342359151510285279856644172747036275008556373211588551217122339164423273129675288374817337956080797215865116453414898513031011920136974314819663282288210631241888549268463460123318118352221488035686636319353484715547646313707068633543465991497138573079552759248473758145439828301035429472888408333215012370553944481950023397524875339764860877259032636492756154172988442461931129